build.assets/tooling/cmd/trigger-workflow/main.go

```diff
@@ -0,0 +1,176 @@
+//     from the raciness of option one, but requires that you have control
+//     of the target workflow.
+//
+// The tool uses the more-generic/more-racy option 1 by default. Run with the
+// `-tag-workflow` option to use the more-specific/less-racy option 2.
+//
+// Note that killing the tool WILL NOT cancel the corresponding GitHub
+// workflow run.
+package main
+
+import (
+	"context"
+	"log"
+	"time"
+
+	ghapi "github.com/google/go-github/v41/github"
+	"github.com/google/uuid"
+	"github.com/gravitational/trace"
+	"golang.org/x/oauth2"
+
+	"github.com/gravitational/teleport/build.assets/tooling/lib/github"
+)
+
+const wokflowTagInput = "workflow-tag"
+
+func main() {
+	args := parseCommandLine()
+	ctx, cancel := context.WithTimeout(context.Background(), args.timeout)
+	defer cancel()
+
+	// Create a GitHub client that authenticates with a Personal Access Token
+	authClient := oauth2.NewClient(ctx, oauth2.StaticTokenSource(
+		&oauth2.Token{AccessToken: args.token},
+	))
+	gh := ghapi.NewClient(authClient)
+
+	dispatchCtx, cancelDispatch := context.WithTimeout(ctx, 1*time.Minute)
+	defer cancelDispatch()
+
+	// Determine what workflows runs have already been created before we start, so
+	// we can exclude them when trying to detect a new run started in response to
+	// our dispatch event. Note that we pick a time slightly in the past to handle
+	// any clock skew.
+	baselineTime := time.Now().Add(-2 * time.Minute)
+	oldRuns, err := github.ListWorkflowRuns(dispatchCtx, gh.Actions, args.owner, args.repo, args.workflow, args.workflowRef, baselineTime)
+	if err != nil {
+		log.Fatalf("Failed to fetch initial task list: %s", err)
+	}
+
+	// If we're tagging the workflow run, then add the tag input to the request so that it
+	// will be piped through to the workflow run
+	var tag string
+	if args.useWorkflowTag {
+		tag = uuid.NewString()
+		args.inputs[wokflowTagInput] = tag
+	}
+
+	log.Printf("Attempting to trigger %s/%s/**/%s @ %s\n", args.owner, args.repo, args.workflow, args.workflowRef)
+	triggerArgs := ghapi.CreateWorkflowDispatchEventRequest{
+		Ref:    args.workflowRef,
+		Inputs: args.inputs,
+	}
+
+	// Issue the workflow_dispatch event.
+	_, err = gh.Actions.CreateWorkflowDispatchEventByFileName(dispatchCtx, args.owner, args.repo, args.workflow, triggerArgs)
+	if err != nil {
+		log.Fatalf("Failed to issue workflow dispatch event: %s", err)
+	}
+
+	run, err := waitForNewWorkflowRun(ctx, gh, args, tag, baselineTime, oldRuns)
+	if err != nil {
+		log.Fatalf("Failed to start workflow run %s", err)
+	}
+	log.Printf("Workflow run: %s", run.GetHTMLURL())
+
+	conclusion, err := github.WaitForRun(ctx, gh.Actions, args.owner, args.repo, args.workflow, args.workflowRef, run.GetID())
+	if err != nil {
+		log.Fatalf("Failed to waiting for run to exit %s", err)
+	}
+
+	if conclusion != "success" {
+		log.Fatalf("Build failed: %s", conclusion)
+	}
+
+	log.Printf("Build succeeded")
```

Comment:
Maybe `Workflow succeeded` instead of `Build succeeded`, for consistent naming with everything else?